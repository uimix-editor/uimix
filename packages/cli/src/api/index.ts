export * from "./context.js";
export * from "./router.js";
