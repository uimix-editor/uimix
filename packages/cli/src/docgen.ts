import * as path from "path";
import { globSync } from "glob";
import docgen from "react-docgen-typescript";
import { Plugin as VitePlugin } from "vite";

export function generateDocs(rootDir: string): docgen.ComponentDoc[] {
  const options: docgen.ParserOptions = {
    savePropValueAsString: true,
    shouldExtractLiteralValuesFromEnum: true,
    componentNameResolver: (exp, source) => {
      //console.log(exp, source);
      return exp.name;
    },
  };

  const filePath = "src/**/*.tsx";
  //const ignoreFilePath = "src/**/*.stories.tsx";

  //const ignoreFilePaths = glob.sync(ignoreFilePath);
  const filePaths = globSync(filePath, {
    cwd: rootDir,
    //ignore: ignoreFilePaths,
  });

  // Parse a file for docgen info
  const docs = docgen
    .withCustomConfig(path.resolve(rootDir, "tsconfig.json"), options)
    .parse(filePaths.map((filePath) => path.resolve(rootDir, filePath)));

  return docs.map((doc) => ({
    ...doc,
    filePath: path.relative(rootDir, doc.filePath),
  }));
}

export function componentsVirtualModulePlugin(rootDir: string): VitePlugin {
  const virtualModuleId = "/virtual:components";
  const resolvedVirtualModuleId = "\0" + virtualModuleId;

  const docs = generateDocs(rootDir);

  const content = `
    import React from "react";
    import ReactDOM from "react-dom/client";
    export const components = ${JSON.stringify(docs)};
    export { React, ReactDOM };
  `;

  return {
    name: "uimix:react-components",
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return content;
      }
    },
    handleHotUpdate({ server }) {
      console.log("TODO: handleHotUpdate");
      return [];
    },
  };
}
