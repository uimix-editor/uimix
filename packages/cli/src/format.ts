import prettier from "prettier/standalone.js";
import parserBabel from "prettier/parser-babel.js";
import parserTypeScript from "prettier/parser-typescript.js";
import parserPostCSS from "prettier/parser-postcss.js";

export function formatJSON(text: string): string {
  return prettier.format(text, {
    parser: "json",
    plugins: [parserBabel],
  });
}

export function formatTypeScript(text: string): string {
  return prettier.format(text, {
    parser: "typescript",
    plugins: [parserTypeScript],
  });
}

export function formatCSS(text: string): string {
  return prettier.format(text, {
    parser: "css",
    plugins: [parserPostCSS],
  });
}
