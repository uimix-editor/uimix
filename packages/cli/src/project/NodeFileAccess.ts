import * as fs from "fs";
import * as path from "path";
import { FileAccess, Stats } from "./FileAccess";
import chokidar from "chokidar";
import { mkdirp } from "mkdirp";
import { globby } from "globby";

export class NodeFileAccess implements FileAccess {
  watch(cwd: string, patterns: string[], onChange: () => void): () => void {
    const watcher = chokidar.watch(`**/{${patterns.join(",")},}`, {
      cwd,
      ignored: ["**/node_modules/**", "**/.git/**"],
    });
    watcher.on("change", onChange);
    watcher.on("add", onChange);
    watcher.on("unlink", onChange);
    return () => watcher.close();
  }

  async glob(cwd: string, patterns: string[]): Promise<string[]> {
    return (
      await globby(`**/{${patterns.join(",")},}`, {
        cwd,
        ignore: ["**/node_modules/**", "**/.git/**"],
      })
    ).map((filePath) => path.resolve(cwd, filePath));
  }

  async stat(filePath: string): Promise<Stats | undefined> {
    try {
      const stat = await fs.promises.stat(filePath);
      return {
        type: stat.isDirectory() ? "directory" : "file",
      };
    } catch (e) {
      return undefined;
    }
  }

  async writeFile(filePath: string, data: Buffer): Promise<void> {
    await mkdirp(path.dirname(filePath));
    await fs.promises.writeFile(filePath, data);
  }

  async readFile(filePath: string): Promise<Buffer> {
    return await fs.promises.readFile(filePath);
  }

  async remove(filePath: string): Promise<void> {
    await fs.promises.rm(filePath);
  }
}
