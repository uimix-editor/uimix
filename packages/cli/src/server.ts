import * as trpcExpress from "@trpc/server/adapters/express";
//import { initSocketIO } from "./socket";
import express from "express";
import cors from "cors";
import { createServer } from "http";
import { renderTrpcPanel } from "trpc-panel";
import { applyWSSHandler } from "@trpc/server/adapters/ws";
import { WebSocketServer } from "ws";
import path from "path";
import * as url from "url";
import { createAppRouter, createContext } from "./api/index.js";
import { createServer as createViteServer } from "vite";
import { componentsVirtualModulePlugin } from "./docgen.js";
import react from "@vitejs/plugin-react";
import { ProjectController } from "./controller/ProjectController.js";
import { Node } from "@uimix/editor/src/models/Node";
console.log(Node);

interface ServerOptions {
  port: number;
  projectPath: string;
}

export async function startServer(options: ServerOptions) {
  const projectController = new ProjectController({
    projectPath: options.projectPath,
  });

  projectController.setupDirectory();
  projectController.generateCode();

  const appRouter = createAppRouter({ projectController });

  const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

  const app = express();
  app.use(
    cors({
      origin: process.env.FRONTEND_URL?.split(","),
    })
  );
  app.use(
    "/trpc",
    trpcExpress.createExpressMiddleware({
      router: appRouter,
      createContext,
    })
  );
  app.use("/panel", (_, res) => {
    return res.send(renderTrpcPanel(appRouter, { url: "/trpc" }));
  });

  if (process.env.NODE_ENV !== "development") {
    app.use(express.static(path.resolve(__dirname, "static")));
  }

  // TODO: make Vite configurable (like Storybook's viteFinal)
  const vite = await createViteServer({
    configFile: false,
    root: options.projectPath,
    server: { middlewareMode: true },
    appType: "custom",
    base: "/project/",
    plugins: [
      componentsVirtualModulePlugin(options.projectPath),
      react({
        exclude: "**/*",
      }),
    ],
  });
  app.use(vite.middlewares);

  const server = createServer(app);
  //initSocketIO(server);

  const wss = new WebSocketServer({
    server,
  });

  applyWSSHandler({
    wss,
    router: appRouter,
  });

  const port = options.port;
  server.listen(port, () => console.log(`listening on port ${port}`));
}
