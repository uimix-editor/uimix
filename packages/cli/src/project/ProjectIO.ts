import * as Data from "@uimix/model/src/data/v1";
import { isEqual } from "lodash-es";
import { formatTypeScript } from "../format";
import { FileAccess } from "./FileAccess";
import * as path from "path";
import { Project } from "@uimix/model/src/models";
import {
  ProjectEmitter,
  ProjectLoader,
  PageNode,
  loadFromJSXFile,
  stringifyAsJSXFile,
  ProjectManifest,
} from "@uimix/model/src/file";
import { getURLSafeBase64Hash } from "@uimix/foundation/src/utils/Hash";
import sizeOf from "image-size";
import { dataUriToBuffer } from "data-uri-to-buffer";
import * as mime from "mime-types";

interface ProjectIOContent {
  manifest: ProjectManifest;
  project: Project;
  pages: Map<string, PageNode>;
  imagePaths: Map<string, string>; // hash to path
}

interface LoadProblem {
  filePath: string;
  error: unknown;
}

interface LoadResult {
  changed: boolean;
  problems: LoadProblem[];
}

// Important TODO: fix paths in Windows!!
export class ProjectIO {
  static readonly projectBoundary = "package.json"; // TODO: other project boundaries

  static async load(fileAccess: FileAccess, rootPath: string) {
    const projectIO = new ProjectIO(fileAccess, rootPath);
    const result = await projectIO.load();
    if (result.problems.length) {
      throw new Error(
        `Problems loading workspace:\n${result.problems
          .map(
            (problem) => `  ${problem.filePath}:\n    ${String(problem.error)}`
          )
          .join("\n")}`
      );
    }
    return projectIO;
  }

  constructor(fileAccess: FileAccess, rootPath: string) {
    this.fileAccess = fileAccess;
    this.rootPath = rootPath;
  }

  readonly fileAccess: FileAccess;
  readonly rootPath: string;

  readonly filePattern = "*.uimix";
  readonly imagePatterns = ["*.png", "*.jpg", "*.jpeg"];
  readonly uimixProjectFile = "uimix.json";

  content: ProjectIOContent = {
    manifest: {},
    project: new Project(),
    pages: new Map(),
    imagePaths: new Map(),
  };

  private async getFilePaths(): Promise<{
    pages: string[];
    images: string[];
  }> {
    // filter out files in sub-projects

    const projectBoundaryFilePaths = await this.fileAccess.glob(this.rootPath, [
      ProjectIO.projectBoundary,
    ]);

    const subProjectPaths = projectBoundaryFilePaths
      .map((filePath) => path.dirname(filePath))
      .filter((projectPath) => this.rootPath !== projectPath);

    const filePaths = (
      await this.fileAccess.glob(this.rootPath, [
        this.filePattern,
        ...this.imagePatterns,
      ])
    ).filter(
      (filePath) =>
        !subProjectPaths.some((projectPath) =>
          filePath.startsWith(projectPath + "/")
        )
    );
    filePaths.sort();

    const pages = filePaths.filter((filePath) => filePath.endsWith(".uimix"));
    const images = filePaths.filter((filePath) => !filePath.endsWith(".uimix"));

    return {
      pages,
      images,
    };
  }

  async load(): Promise<LoadResult> {
    const problems: LoadProblem[] = [];

    try {
      const filePaths = await this.getFilePaths();

      let manifest: ProjectManifest = {};

      const manifestPath = path.join(this.rootPath, this.uimixProjectFile);
      if ((await this.fileAccess.stat(manifestPath))?.type === "file") {
        try {
          manifest = ProjectManifest.parse(
            JSON.parse(
              (
                await this.fileAccess.readFile(
                  path.join(this.rootPath, this.uimixProjectFile)
                )
              ).toString()
            )
          );
        } catch (error) {
          problems.push({
            filePath: manifestPath,
            error,
          });
        }
      }

      const images = new Map<string, Data.Image>();
      const pages = new Map<string, PageNode>();
      const imagePaths = new Map<string, string>();

      // TODO: reload changed files only
      for (const filePath of filePaths.pages) {
        try {
          const pageText = (
            await this.fileAccess.readFile(filePath)
          ).toString();
          const pageNode = loadFromJSXFile(pageText);
          const pageName = path
            .relative(this.rootPath, filePath)
            .replace(/\.uimix$/, "");
          pages.set(pageName, pageNode);
        } catch (error) {
          problems.push({
            filePath,
            error,
          });
        }
      }

      for (const filePath of filePaths.images) {
        try {
          // TODO: lookup specific directories only
          const imageData = await this.fileAccess.readFile(filePath);
          const hash = await getURLSafeBase64Hash(imageData);
          const mimeType = mime.lookup(filePath) || "image/png";
          const size = sizeOf(imageData);

          const image: Data.Image = {
            width: size.width ?? 0,
            height: size.height ?? 0,
            type: mimeType as Data.ImageType,
            url: `data:${mimeType};base64,${imageData.toString("base64")}`,
          };
          images.set(hash, image);
          imagePaths.set(hash, filePath);
        } catch (error) {
          problems.push({
            filePath,
            error,
          });
        }
      }

      const imagePathToHash = new Map<string, string>();
      for (const [hash, absPath] of imagePaths) {
        imagePathToHash.set(path.relative(this.rootPath, absPath), hash);
      }

      const loader = new ProjectLoader(this.content.project, imagePathToHash);
      loader.load(pages);
      for (const [key, image] of images) {
        loader.project.imageManager.images.set(key, image);
      }

      if (isEqual(pages, this.content.pages)) {
        return {
          changed: false,
          problems,
        };
      }

      this.content.manifest = manifest;
      this.content.pages = pages;
      this.content.imagePaths = imagePaths;

      return {
        changed: true,
        problems,
      };
    } catch (error) {
      problems.push({
        filePath: this.rootPath,
        error,
      });

      return {
        changed: false,
        problems,
      };
    }
  }

  private isSaving = false;

  async save(): Promise<void> {
    try {
      this.isSaving = true;
      const project = this.content.project;

      // save images

      const imagePathsInProject = new Map<string, string>();

      for (const imagePath of (await this.getFilePaths()).images) {
        const imageData = await this.fileAccess.readFile(imagePath);
        const hash = await getURLSafeBase64Hash(imageData);
        imagePathsInProject.set(hash, path.relative(this.rootPath, imagePath));
      }

      const usedImageHashes = project.imageManager.usedImageHashes;

      for (const [hash, image] of project.imageManager.images) {
        if (!usedImageHashes.has(hash)) {
          // not used
          continue;
        }
        if (imagePathsInProject.has(hash)) {
          // already saved
          continue;
        }

        const decoded = dataUriToBuffer(image.url);
        const suffix = mime.extension(decoded.type) || "bin";
        await this.fileAccess.writeFile(
          // TODO: make default image directory configurable
          path.join(this.rootPath, "src/images", `${hash}.${suffix}`),
          decoded
        );
      }

      // save pages

      const pagePathsToDelete = new Set(
        await this.fileAccess.glob(this.rootPath, [this.filePattern])
      );

      const projectEmitter = new ProjectEmitter(project, imagePathsInProject);
      const pages = projectEmitter.emit();

      for (const [pageName, pageNode] of pages) {
        const pagePath = path.join(this.rootPath, pageName + ".uimix");
        await this.fileAccess.writeFile(
          pagePath,
          Buffer.from(formatTypeScript(stringifyAsJSXFile(pageNode)))
        );
        pagePathsToDelete.delete(pagePath);
      }

      // TODO: save manifest

      // const manifestPath = path.join(projectPath, this.uimixProjectFile);

      // let parsed: ProjectManifestJSON;
      // try {
      //   parsed = JSON.parse(
      //     (await this.fileAccess.readFile(manifestPath)).toString()
      //   ) as ProjectManifestJSON;
      // } catch (e) {
      //   parsed = {};
      // }
      // parsed.prebuiltAssets = project.manifest.prebuiltAssets;

      // TODO: avoid overwriting malformed uimix.json
      // await this.fileAccess.writeFile(
      //   manifestPath,
      //   Buffer.from(formatJSON(JSON.stringify(parsed)))
      // );

      this.content.pages = pages;

      for (const pagePath of pagePathsToDelete) {
        await this.fileAccess.remove(pagePath);
      }
    } finally {
      this.isSaving = false;
    }
  }

  watch(onChange: () => void): () => void {
    console.log("start watching...");

    return this.fileAccess.watch(
      this.rootPath,
      [this.filePattern, ProjectIO.projectBoundary],
      async () => {
        try {
          if (this.isSaving) {
            return;
          }
          if ((await this.load()).changed) {
            onChange();
          }
        } catch (e) {
          console.error(e);
        }
      }
    );
  }
}
