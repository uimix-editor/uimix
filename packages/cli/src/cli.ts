import { cac } from "cac";
import { startServer } from "./server.js";

const cli = cac("uimix");

cli
  .command("[root]", "start the editor server")
  .option("--port <port>", `[number] specify port`)
  .action(
    async (
      root: string | undefined,
      options: {
        port?: number;
      }
    ) => {
      startServer({
        projectPath: root ?? ".",
        port: options.port ?? 4000,
      });
    }
  );

cli.command("build [root]", "build the UIMix files").action(() => {
  console.log("TODO");
});

cli.help();
cli.version("0.0.1");
cli.parse();
