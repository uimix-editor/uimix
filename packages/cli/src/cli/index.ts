import { cac } from "cac";
import { generateCode } from "../compiler/generateCode";
import * as fs from "fs";
import * as path from "path";
import mkdirp from "mkdirp";
import { ProjectIO } from "../project/ProjectIO";
import { NodeFileAccess } from "../project/NodeFileAccess";
import { buildCodeAssets } from "../codeAssets/build";

async function compileProject(projectIO: ProjectIO) {
  const outFiles = await generateCode(
    projectIO.rootPath,
    projectIO.content.manifest,
    projectIO.content.project.toJSON(),
    projectIO.content.imagePaths
  );

  for (const outFile of outFiles) {
    const outPath = path.join(projectIO.rootPath, outFile.filePath);
    const outPathDir = path.dirname(outPath);
    mkdirp.sync(outPathDir);
    fs.writeFileSync(outPath, outFile.content);
  }
}

async function compileCommand(
  rootPath: string,
  options: {
    watch?: boolean;
  }
): Promise<void> {
  const projectIO = await ProjectIO.load(new NodeFileAccess(), rootPath);

  if (options.watch) {
    projectIO.watch(() => compileProject(projectIO));
  }

  await buildCodeAssets(rootPath, projectIO.content.manifest, options);
  await compileProject(projectIO);
}

const cli = cac("uimix");

// TODO: output help if unknown command is passed

cli
  .command("compile", "compile UIMix project")
  .option("--root", `[string] root directory`)
  .option("-w, --watch", `[boolean] watch for changes`)
  .action(async (options: { root?: string; watch?: boolean }) => {
    await compileCommand(path.resolve(options.root || "."), options);
  });

cli.help();
cli.version("0.0.1");
cli.parse();
