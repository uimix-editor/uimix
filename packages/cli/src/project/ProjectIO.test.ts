import { describe, it, expect, beforeEach, afterEach } from "vitest";
import * as fs from "fs";
import shell from "shelljs";
import tmp from "tmp";
import { Project } from "@uimix/model/src/models";
import { ProjectEmitter } from "@uimix/model/src/file";
import { NodeFileAccess } from "./NodeFileAccess";
import { ProjectIO } from "./ProjectIO";

describe(ProjectIO.name, () => {
  let tmpObj: tmp.DirResult;

  beforeEach(async () => {
    tmpObj = tmp.dirSync({
      unsafeCleanup: true,
    });
    shell.mkdir("-p", tmpObj.name + "/demo-project");
  });

  afterEach(async () => {
    tmpObj.removeCallback();
  });

  // it("loads", () => {
  //   const projectFiles = new ProjectFiles(tmpObj.name + "/demo-project");
  //   projectFiles.load();

  //   expect(projectFiles.toProjectJSON()).toMatchSnapshot();
  // });

  it("saves", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const projectIO = new ProjectIO(
      new NodeFileAccess(),
      tmpObj.name + "/demo-project"
    );
    projectIO.content.project = project;
    await projectIO.save();

    const page1File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page1.uimix",
      "utf8"
    );
    expect(page1File).toMatchSnapshot();

    const page2File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page2.uimix",
      "utf8"
    );
    expect(page2File).toMatchSnapshot();

    const projectIO2 = await ProjectIO.load(
      new NodeFileAccess(),
      tmpObj.name + "/demo-project"
    );

    expect(
      new ProjectEmitter(projectIO2.content.project, new Map()).emit()
    ).toEqual(new ProjectEmitter(project, new Map()).emit());
  });

  it("watches project", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const projectIO = new ProjectIO(
      new NodeFileAccess(),
      tmpObj.name + "/demo-project"
    );
    projectIO.content.project = project;

    await projectIO.save();
    await new Promise((resolve) => setTimeout(resolve, 500));

    let watchCount = 0;
    projectIO.watch(() => {
      watchCount++;
    });
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(0);

    console.log("save");
    // saves not cause watch
    await projectIO.save();
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(0);

    console.log("change");
    // change file
    fs.rmSync(tmpObj.name + "/demo-project/src/page1.uimix");
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(1);

    expect(
      projectIO.content.project.pages.all.map((page) => page.filePath)
    ).toEqual(["src/page2"]);
  });
});
