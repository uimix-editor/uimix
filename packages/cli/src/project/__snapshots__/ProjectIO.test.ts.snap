// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectIO > saves 1`] = `
"export default (
  <page>
    <frame name=\\"\\" />
  </page>
);
"
`;

exports[`ProjectIO > saves 2`] = `
"export default (
  <page>
    <text name=\\"\\" />
  </page>
);
"
`;
